File: android/hardware/aw/atw/tests/py_test/atw_float2int.py
# -*- coding: utf-8 -*-
import sys,os,math
import numpy as np

def GetShiftBits(tmp):
    flag = 0;
    if(tmp<1.0):
        flag = 0;
    elif(tmp<2.0):
        flag = 1;
    elif(tmp<4.0):
        flag = 2;
    elif(tmp<8.0):
        flag = 3;
    elif(tmp<16.0):
        flag = 4;
    elif(tmp<32.0):
        flag = 5;
    elif(tmp<64.0):
        flag = 6;
    elif(tmp<128.0):
        flag = 7;
    elif(tmp<256.0):
        flag = 8;
    elif(tmp<512.0):
        flag = 9;
    elif(tmp<1024.0):
        flag = 10;
    elif(tmp<2048.0):
        flag = 11;
    else:
        flag = 12;
    return flag;

def GetShift(ouw, ouh):
    shift = 0;
    if(ouw >= 0x800):
        shift = 2;
    elif(ouw >= 0x400):
        shift = 3;
    elif(ouw >= 0x200):
        shift = 4;
    else:
        shift = 8;

    tmpshift = shift;
    if(ouh >= 0x800):
        shift = 2;
    elif(ouh >= 0x400):
        shift = 3;
    elif(ouh >= 0x200):
        shift = 4;
    else:
        shift = 8;

    if(shift < tmpshift):
        shift = tmpshift;
    return shift;

def ConvertFromFloat2Int(wp, hp, flag, shift, fdata, log):
    data = np.array(fdata);
    ret_data = np.arange(8, dtype='int32');
    for i in range(0,8):
        tmp = data[i];
        if flag > 0:
            tmp = tmp / (1<<flag);
        if i == 0 or i == 3 or i == 6:
            tmp = tmp * hp;
        else:
            tmp = tmp * wp;
        ret_data[i] = tmp * (1<<shift);
        if log == 1:
            print (hp, wp);
            print ("for: i=",i ,", int=" , ret_data[i] , ", float=", data[i] , ", tmp=" , tmp);
    return ret_data;

File: android/hardware/aw/atw/tests/test_app/AtwCoefficentsCaculator/src/atw_float2int.cpp
#include <stdlib.h>
#include <stdint.h>
#include <math.h>

static double GetMaxOfArray(double p[])//TODO: 求8个浮点数中的最大绝对值
{
        double tmp = fabs(p[0]);
        for(int i = 1; i < 8; i++){
                tmp = tmp < fabs(p[i]) ? tmp : fabs(p[i]);
        }
        return tmp;
}

static size_t GetShiftBits(double tmp)
{
        size_t flag = 0;
        if(tmp<1.0){
                flag = 0;
        }
        else if(tmp<2.0){
                flag = 1;
        }
        else if(tmp<4.0){
                flag = 2;
        }
        else if(tmp<8.0){
                flag = 3;
        }
        else if(tmp<16.0){
                flag = 4;
        }
        else if(tmp<32.0){
                flag = 5;
        }
        else if(tmp<64.0){
                flag = 6;
        }
        else if(tmp<128.0){
                flag = 7;
        }
        else if(tmp<256.0){
                flag = 8;
        }
        else if(tmp<512.0){
                flag = 9;
        }
        else if(tmp<1024.0){
                flag = 10;
        }
        else if(tmp<2048.0){
                flag = 11;
        }
        else{
                flag = 12;
        }
        return flag;
}

unsigned char GetShift(int ouw, int ouh)
{
    unsigned char shift = 0;
    if(ouw >= 0x800)
        shift = 2;
    else if(ouw >= 0x400)
        shift = 3;
    else if(ouw >= 0x200)
        shift = 4;
    else
        shift = 8;
    unsigned char tmpshift = shift;
    if(ouh >= 0x800)
        shift = 2;
    else if(ouh >= 0x400)
        shift = 3;
    else if(ouh >= 0x200)
        shift = 4;
    else
        shift = 8;
    if(shift < tmpshift)
        shift = tmpshift;
    return shift;
}

size_t ConvertFromFloat2Int(int wp, int hp, unsigned char shift, double floatCoeffients[], int32_t outputIntCoeffients[])
{
    double maxd = GetMaxOfArray(floatCoeffients);
    size_t flag = GetShiftBits(maxd);
    for(int i=0; i<8; i++)
    {
        double tmp = floatCoeffients[i];
        if(flag)
        {
            tmp = tmp/(1<<flag);
        }
        switch(i)
        {
            case 0:
            case 3:
            case 6:
                tmp = tmp*hp;
                break;
            default:
                tmp = tmp*wp;
                break;
        }
        outputIntCoeffients[i] = (int32_t)(tmp*(1<<shift));
    }
    return flag;
}
